src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -1514,37 +1536,53 @@ class Simplex_tree {
   }
 
  public:
-  /** \brief This function ensures that each simplex has a higher filtration value than its faces by increasing the
-   * filtration values.
-   * @return True if any filtration value was modified, false if the filtration was already non-decreasing.
-   * 
-   * If a simplex has a `NaN` filtration value, it is considered lower than any other defined filtration value.
+  /** Calls a function on each simplex. The order ensures that faces are visited before cofaces.
+   * While it is fine to modify the data of a simplex (filtration, key) in the function, modifying
+   * the structure itself (insertion, removal) is not supported.
+   *
+   * @param[in] fun Function that takes as argument a Simplex_handle and an int (representing the dimension of this
+   * simplex). It may return void or bool, and in the second case returning true means that the iteration will skip
+   * the children of this simplex (a subset of the cofaces).
    */
-  bool make_filtration_non_decreasing() {
-    bool modified = false;
-    // Loop must be from the end to the beginning, as higher dimension simplex are always on the left part of the tree
-    for (auto& simplex : boost::adaptors::reverse(root_.members())) {
-      if (has_children(&simplex)) {
-        modified |= rec_make_filtration_non_decreasing(simplex.second.children());
+  template<class Fun>
+  void for_each_simplex(Fun&& fun) {
+    auto f=[&fun](Simplex_handle sh, int dim){
+      if constexpr (std::is_same_v<void, decltype(fun(sh, dim))>) {
+        fun(sh, dim);
+        return false;
+      } else {
+        return fun(sh, dim);
       }
-    }
-    if(modified)
-      clear_filtration(); // Drop the cache.
-    return modified;
+    };
+    if (!is_empty())
+      for_each_simplex(root(), 0, f);
   }
 
  private:
-  /** \brief Recursively Browse the simplex tree to ensure the filtration is not decreasing.
-   * @param[in] sib Siblings to be parsed.
-   * @return The filtration modification information in order to trigger initialize_filtration.
+  template<class Fun>
+  void for_each_simplex(Siblings* sib, int dim, Fun&& fun) {
+    for (auto& simplex : boost::adaptors::reverse(sib->members())) {
+      Simplex_handle sh(&simplex);
+      if (!fun(sh, dim) && has_children(sh)) {
+        for_each_simplex(sh->second.children(), dim+1, fun);
+      }
+      // We could skip checking has_children for the first element of the iteration, we know it returns false.
+    }
+  }
+
+ public:
+  /** \brief This function ensures that each simplex has a higher filtration value than its faces by increasing the
+   * filtration values.
+   * @return True if any filtration value was modified, false if the filtration was already non-decreasing.
+   *
+   * If a simplex has a `NaN` filtration value, it is considered lower than any other defined filtration value.
    */
-  bool rec_make_filtration_non_decreasing(Siblings * sib) {
+  bool make_filtration_non_decreasing() {
```

Comment:
Is the (small) slowdown reproducible if you run the benchmark several times? The benchmark I tried was too noisy to see the difference (say I would get .085 vs .082 like you, but the next run would give .09 for master...).

Reply:
Ok, now I got it. It was because working on the simplex tree copy was faster than the original one (as you explained it to me). When working on 2 copies and whatever the order the methods are launched, I have now:
```txt
$ ./simplex_tree_foreach_benchmark 40000
** Without cofaces computation optimization
... Simplex_tree construction
... 38844 vertices and 459036 simplices.
Benchmark make_filtration_non_decreasing from PR: 0.057s

Benchmark make_filtration_non_decreasing from master: 0.061s

** With cofaces computation optimization
... Simplex_tree construction
... 38825 vertices and 457719 simplices.
Benchmark make_filtration_non_decreasing from PR: 0.066s

Benchmark make_filtration_non_decreasing from master: 0.07s
```